src/main/java/com/celements/ressource_url/RessourceUrlService.java

```diff
@@ -0,0 +1,185 @@
+  }
+
+  @Override
+  @NotEmpty
+  public String getRessourceURLPrefix(@NotEmpty String action) {
+    XWikiURLFactory urlf = context.getXWikiContext().getURLFactory();
+    return urlf.createResourceURL("", true, context.getXWikiContext()).toString().replace("/skin/",
+        "/" + action + "/");
+  }
+
+  @Override
+  @NotNull
+  public String getExternalRessourceURL(@NotNull String fileName,
+      @NotNull Optional<String> action) {
+    try {
+      return context.getXWikiContext().getURLFactory().getServerURL(context.getXWikiContext())
+          .toExternalForm() + createRessourceUrl(fileName, action);
+    } catch (MalformedURLException | UrlRessourceNotExistException exp) {
+      LOGGER.error("Failed to getServerURL.", exp);
+    }
+    return "";
+  }
+
+  @Override
+  @NotNull
+  public String createRessourceUrl(@NotNull String link, @NotNull Optional<String> action,
+      @NotNull Optional<String> queryString) throws UrlRessourceNotExistException {
+    final String baseUrl = createRessourceUrl(link, action);
+    if (queryString.isPresent()) {
+      if (baseUrl.indexOf("?") > -1) {
+        return baseUrl + "&" + queryString.get();
+      } else {
+        return baseUrl + "?" + queryString.get();
+      }
+    }
+    return baseUrl;
+  }
+
+  @Override
+  public @NotEmpty String createRessourceUrl(@NotNull String link, @NotNull Optional<String> action)
+      throws UrlRessourceNotExistException {
+    String url = link;
+    if (isAttachmentLink(link)) {
+      url = createAttachmentUrl(link, action);
+    } else if (isOnDiskLink(link)) {
+      url = createOnDiskUrl(link, action);
+    }
+    return addContextUrl(url);
+  }
+
+  private String addContextUrl(String url) {
+    Optional<XWikiDocument> currentDoc = context.getCurrentDoc().toJavaUtil();
+    if (currentDoc.isPresent() && url.startsWith("?")) {
+      url = currentDoc.get().getURL("view", context.getXWikiContext()) + url;
+    }
+    return url;
+  }
+
+  private String createAttachmentUrl(String link, Optional<String> action)
+      throws UrlRessourceNotExistException {
+    String attName = getAttachmentName(link);
+    try {
+      XWikiDocument doc = modelAccess.getDocument(getPageDocRef(link));
+      XWikiAttachment att = attachmentSrv.getAttachmentNameEqual(doc, attName);
+      return doc.getAttachmentURL(attName, getAction(action), context.getXWikiContext())
+          + "?version=" + lastStartupTimeStamp.getLastChangedTimeStamp(att.getDate());
+    } catch (DocumentNotExistsException exp) {
+      LOGGER.error("Error getting attachment URL for doc '{}' and file {}", getPageDocRef(link),
+          attName, exp);
+      throw new UrlRessourceNotExistException(link);
+    } catch (AttachmentNotExistsException anee) {
+      LOGGER.info("Attachment not found for link [{}] and action [{}]", link, action, anee);
+      throw new UrlRessourceNotExistException(link);
+    }
+  }
+
+  private String createOnDiskUrl(String link, Optional<String> action) {

```

Comment:
for all methods ending with Url, better return an actual `java.net.URL` instance. see (and perhaps use) `com.celements.web.service.UrlService.createURIBuilder(EntityReference)` for building URI/URLs